### ISSUE TITLE
【Performance】fetchall performance 

### ISSUE THREAD

**Issue by WildGooseSouth** [PyGreSQL/PyGreSQL#51]
Hi，recently we encounter an performance problem when we upgrade the PyGreSQL from 4.1 to 5.1 with Python2.7 runing environment.

The test case is just consider about the fetchall interface. It's just query 200M rows. The test case is 

```
import datetime
import time
import pgdb
try:
    conn=pgdb.connect(host='127.0.0.1:5433',user='postgres',database='postgres')
    c=conn.cursor()
    c.execute("select * from XXX_int")
    row = c.fetchall()
    c.close()
    conn.close()
except (pgdb.InternalError,Exception) as e:
    print(str(e) + "\nFAILED")

```
The server version is PostgreSQL12.4, it's information as follows
```
postgres=# select version();
                                                 version
---------------------------------------------------------------------------------------------------------
 PostgreSQL 12.4 on x86_64-pc-linux-gnu, compiled by gcc (GCC) 4.8.5 20150623 (EulerOS 4.8.5-28), 32-bit
(1 row)

postgres=# select count(1) from XXX_int;
  count
---------
 2000000
(1 row)

postgres=# \d xxx_int
              Table "public.xxx_int"
 Column |  Type   | Collation | Nullable | Default
--------+---------+-----------+----------+---------
 a      | integer |           |          |
 b      | integer |           |          |
```
For private reason, we need to compile 32-bit version in X86_64 platform with the -m32 flag.

We compare the execution time with two version.

py4.1
```
[/PyGreSQL-4.1/module/build/lib.linux-x86_64-2.7]$ time python test.py

real    0m7.868s
user    0m7.139s
sys     0m0.325s
[/PyGreSQL-4.1/module/build/lib.linux-x86_64-2.7]$ time python test.py

real    0m8.237s
user    0m7.473s
sys     0m0.333s
[/PyGreSQL-4.1/module/build/lib.linux-x86_64-2.7]$ time python test.py

real    0m7.856s
user    0m7.151s
sys     0m0.298s
```
py5.1
```
[/usr/local/lib/python2.7/site-packages/PyGreSQL-5.1-py2.7-linux-x86_64.egg]$ time python test.py

real    0m13.153s
user    0m12.408s
sys     0m0.318s
[/usr/local/lib/python2.7/site-packages/PyGreSQL-5.1-py2.7-linux-x86_64.egg]$ time python test.py

real    0m13.223s
user    0m12.505s
sys     0m0.291s
[/usr/local/lib/python2.7/site-packages/PyGreSQL-5.1-py2.7-linux-x86_64.egg]$ time python test.py

real    0m13.261s
user    0m12.535s
sys     0m0.313s
```
You can see the avg time is about 8s in py4.1 while 13.2s in py5.1.

【Analyze】
Trough the analyze, we find the most cost time is in the following at the fetchmany function end, which is used for constructing the returning result.

```
return [row_factory([typecast(value, typ)
            for typ, value in zip(coltypes, row)]) for row in result]
```
Comparing the fetchmany function, we find a little differences and rewrite it.

```
typecast = self.type_cache.typecast
row_factory = self.row_factory
coltypes = self.coltypes
return [row_factory([typecast(value, typ)
            for typ, value in zip(coltypes, row)]) for row in result]
```
The executing time is as follows.
```
[ /usr/local/lib/python2.7/site-packages/PyGreSQL-5.1-py2.7-linux-x86_64.egg]$ time python test.py

real    0m11.158s
user    0m10.451s
sys     0m0.288s
[ /usr/local/lib/python2.7/site-packages/PyGreSQL-5.1-py2.7-linux-x86_64.egg]$ time python test.py

real    0m11.226s
user    0m10.530s
sys     0m0.300s
```
Through the testing result, execution time is down from 13s to 11s.  

And another difference is funcion **typecast**， but I donnot know how to optimiation it.

**Comment by Cito** [comment]
Thanks for reporting, @WildGooseSouth.

Your optimization in `fetchmany`, avoiding the attribute lookup inside the list comprehension, makes total sense, and I can imagine that it's measurable if you have that many rows. I have already committed this change.

It would be also interesting to check whether passing the argument to row_factory as list, tuple or generator also makes a difference (this should be checked in Python 2.8 and 3.9 because they could behave differently). Will experiment a little bit when I have more time, maybe you want to do some experimentation as well.

list:
```python3
       return [row_factory([typecast(value, typ)
                for typ, value in zip(coltypes, row)]) for row in result]
```

tuple:
```python3
       return [row_factory(tuple(typecast(value, typ)
                for typ, value in zip(coltypes, row))) for row in result]
```

generator:
```python3
       return [row_factory(typecast(value, typ)
                for typ, value in zip(coltypes, row)) for row in result]
```

**Comment by Cito** [comment]
Here is another possible optimization that should be checked:

```python3
        typecasts = [partial(typecast, typ=typ) for typ in self.coltypes]
        return [row_factory(typecast(value)
                for typecast, value in zip(typecasts, row)) for row in result]
```

But I have not tested if this really is faster or maybe even slower.

Regarding the typecast, it is now more flexible and extensible than in 4.1, but of course this comes with a certain cost. We could try to optimize the fetching of the proper cast function and/or the cast functions themselves.

The crucial problem is probably that there still is a `cast = self.get_typecast(typ)` lookup per value inside the `typecast` method. This could theoretically be optimized. The `TypeCache` class could have another method that does not take a `typ` parameter, but a `cast` parameter, and these parameters could be prefetched in `fetchmany` as shown above.

**Comment by WildGooseSouth** [comment]
Yes, the data count is 2000000. The difference is about 

I test the new modification with Python3.8.2 .   Comparing with the no modification, it seems too worse.

```
        typecast = self.type_cache.typecast
        row_factory = self.row_factory
        coltypes = self.coltypes
        return [row_factory([typecast(value, typ)
            for typ, value in zip(coltypes, row)]) for row in result]
```
Testing time is
```
[PyGreSQL-5.1/build/lib.linux-x86_64-3.8]$ time python3 test.py

real    0m10.385s
user    0m9.714s
sys     0m0.252s
[PyGreSQL-5.1/build/lib.linux-x86_64-3.8]$ time python3 test.py

real    0m10.376s
user    0m9.677s
sys     0m0.274s
[ PyGreSQL-5.1/build/lib.linux-x86_64-3.8]$ time python3 test.py

real    0m10.441s
user    0m9.751s
sys     0m0.254s
```
With new modification, the execution time become long.
```
[PyGreSQL-5.1/build/lib.linux-x86_64-3.8]$ time python3 test.py

real    0m13.163s
user    0m12.452s
sys     0m0.279s
[PyGreSQL-5.1/build/lib.linux-x86_64-3.8]$ time python3 test.py

real    0m13.334s
user    0m12.609s
sys     0m0.287s
[PyGreSQL-5.1/build/lib.linux-x86_64-3.8]$ time python3 test.py

real    0m12.755s
user    0m12.061s
sys     0m0.274s
```
the code is
```
        typecast = self.type_cache.typecast
        row_factory = self.row_factory
        coltypes = self.coltypes
        typecasts = [partial(typecast, typ=typ) for typ in self.coltypes]
        return [row_factory(typecast(value)
                for typecast, value in zip(typecasts, row)) for row in result]
```
When I print the typecasts variable, the result is
```
[functools.partial(<bound method TypeCache.typecast of {23: 'int4', 'int4': 'int4'}>, typ='int4'), functools.partial(<bound method TypeCache.typecast of {23: 'int4', 'int4': 'int4'}>, typ='int4')]
```
And I think you hope to get the result like this
```
typecasts = [int, int]
```
If I use the hard-code for typecasts, the execution time is smaller.
```
[PyGreSQL-5.1/build/lib.linux-x86_64-3.8]$ time python3 test.py

real    0m7.662s
user    0m6.998s
sys     0m0.265s
````

**Comment by Cito** [comment]
Ok, thanks for testing. I leave this open for now and will look into this again later.

**Comment by WildGooseSouth** [comment]
I test the psycopg2 and find the it's execute too fast. Read the psycopg2 souce code, I find two differences.

1、The cast function in C not in the python.

From the code （https://github.com/psycopg/psycopg2/blob/master/lib/extras.py), we can see the fetchall

```
    def fetchall(self):
        ts = super(NamedTupleCursor, self).fetchall()
        nt = self.Record
        if nt is None:
            nt = self.Record = self._make_nt()
        return list(map(nt._make, ts))
```
From the C code (https://github.com/psycopg/psycopg2/blob/master/psycopg/typecast_basic.c), we can see the cast function.

And in the function _psyco_curs_buildrow_fill,  we see the typecast_cast function

```
PyObject *
typecast_cast(PyObject *obj, const char *str, Py_ssize_t len, PyObject *curs)
{
    PyObject *old, *res = NULL;
    typecastObject *self = (typecastObject *)obj;

    Py_INCREF(obj);
    old = ((cursorObject*)curs)->caster;
    ((cursorObject*)curs)->caster = obj;

    if (self->ccast) {
        res = self->ccast(str, len, curs);
    }
    else if (self->pcast) {
        PyObject *s;
        /* XXX we have bytes in the adapters and strings in the typecasters.
         * are you sure this is ok?
         * Notice that this way it is about impossible to create a python
         * typecaster on a binary type. */
        if (str) {
#if PY_2
            s = PyString_FromStringAndSize(str, len);
#else
            s = conn_decode(((cursorObject *)curs)->conn, str, len);
#endif
        }
        else {
            Py_INCREF(Py_None);
            s = Py_None;
        }
        if (s) {
            res = PyObject_CallFunctionObjArgs(self->pcast, s, curs, NULL);
            Py_DECREF(s);
        }
    }
    else {
        PyErr_SetString(Error, "internal error: no casting function found");
    }

    ((cursorObject*)curs)->caster = old;
    Py_DECREF(obj);

    return res;
}
```
It use the  self->ccast to cast node. For example, convert int to PythonInt
```
/** LONGINTEGER - cast long integers (8 bytes) to python long **/

static PyObject *
typecast_LONGINTEGER_cast(const char *s, Py_ssize_t len, PyObject *curs)
{
    char buffer[24];

    if (s == NULL) { Py_RETURN_NONE; }
    if (s[len] != '\0') {
        strncpy(buffer, s, (size_t) len); buffer[len] = '\0';
        s = buffer;
    }
    return PyLong_FromString((char *)s, NULL, 0);
}
```
So I try use this function in the source_fetch
```
#if IS_PY3
                if (PQfformat(self->result, j) == 0) { /* textual format */
                    str = get_decoded_string(s, size, encoding);
                    if (!str) /* cannot decode */
                        str = PyBytes_FromStringAndSize(s, size);
                }
                else
#endif
                if (PQftype(self->result, j) == 23) {
                   str = typecast_INTEGER_cast(s, size, NULL);
                } else {
                   str = PyBytes_FromStringAndSize(s, size);
                }
```
Also, in pgdb.py, I change the return as
```
return [row_factory([value for typ, value in zip(coltypes, row)]) for row in result]
```
The executing time is
```
[PyGreSQL-5.1/build/lib.linux-x86_64-2.7]$ time python test.py

real    0m5.307s
user    0m4.692s
sys     0m0.230s
[PyGreSQL-5.1/build/lib.linux-x86_64-2.7]$ time python test.py

real    0m5.221s
user    0m4.587s
sys     0m0.214s
[PyGreSQL-5.1/build/lib.linux-x86_64-2.7]$ time python test.py

real    0m5.234s
user    0m4.608s
sys     0m0.188s
```
2、I find the return value in pygresql is like [ROW(XXX),ROW(XXXX)], but in the psycopg2, it's just a list donnot contain ROW information.  Do we need to change the return format?

**Comment by Cito** [comment]
The row format in PyGreSQL is flexible and can be customized using a row factory. By detault it returns named tuples which are more convenient than normal tuples. But you can also make it return dicts or normal tuples (see docstrings/documentation). Psycopg2 returns normal tuples by default, but the extra package also supports named tuples (you are quoting code from that package). Of course normal tuples are faster if you create millions of them, so maybe you want to use a custom row factory to optimize for performance.

The other issue is type casting. Note that the heavy typecasting functions like for arrays and records are written in C code in PyGreSQL, too. The problem here is mainly with the typecast lookups. Currently this happens for every single value, based on the type and on the value (e.g. null values are alsways returned as None). This can of course be optimized if you have many rows, by determining the typecast functions only once, since they only depend on the column, not on the row.

**Comment by Cito** [comment]
I have now commited two changes. The last one particularly optimizes the case of fetching many rows. Together I measure about 40% performance boost. Can you try out the master branch and let me know how it works for you?

If you want to speed it up even more, you can use raw tuples instead of named tuples:

```python3
class TupleCursor(pgdb.Cursor):

    row_factory = staticmethod(tuple)
```

Then get your cursor with `TupleCursor(con)` instead of `con.cursor()` or set `con.cursor_type = TupleCursor` to always return tuples for this connection.

**Comment by Cito** [comment]
The optimizations are now available in version 5.2.2.

**Comment by WildGooseSouth** [comment]
This two weeks I've too many things to do.  I'm sorry for testint it later. The testing reuslt in my env is
```
[/PyGreSQL-5.1/build/lib.linux-x86_64-2.7]$ time python test.py

real    0m6.335s
user    0m5.727s
sys     0m0.185s
[/PyGreSQL-5.1/build/lib.linux-x86_64-2.7]$ time python test.py

real    0m6.514s
user    0m5.899s
sys     0m0.185s
[/PyGreSQL-5.1/build/lib.linux-x86_64-2.7]$ time python test.py

real    0m6.235s
user    0m5.604s
sys     0m0.200s
```

The result is better than the old version. And my customer tell me the performance also better than old version from the product view. Thanks for your helping again.
